File: src/accessiweather/surf_conditions.py
"""Surf and beach-condition helpers for official and derived sources."""

from __future__ import annotations

import inspect
import logging
from dataclasses import dataclass
from datetime import UTC, datetime
from typing import Any

import httpx

from .models import Location, TextProduct
from .weather_client_parsers import degrees_to_cardinal

logger = logging.getLogger(__name__)

OPENMETEO_MARINE_BASE_URL = "https://marine-api.open-meteo.com/v1"
_OPENMETEO_CURRENT_VARIABLES = (
    "wave_height",
    "wave_direction",
    "wave_period",
    "swell_wave_height",
    "swell_wave_direction",
    "swell_wave_period",
    "sea_surface_temperature",
)


@dataclass(frozen=True)
class SurfConditionReport:
    """A source-labelled surf/beach conditions summary."""

    source_name: str
    product_id: str
    text: str
    issued_at: datetime | None = None
    official: bool = False

    def to_text_product(self) -> TextProduct:
        """Convert the report to the TextProduct shape used by Forecaster Notes."""
        return TextProduct(
            product_type="SURF_CONDITIONS",
            product_id=self.product_id,
            cwa_office=self.source_name,
            issuance_time=self.issued_at,
            product_text=self.text,
            headline=(
                f"Surf conditions from {self.source_name}"
                if not self.official
                else f"Official surf forecast from {self.source_name}"
            ),
        )


async def _client_get(
    client: httpx.AsyncClient,
    url: str,
    *,
    params: dict[str, Any] | None = None,
    headers: dict[str, str] | None = None,
) -> httpx.Response:
    """Call AsyncClient.get while tolerating synchronous test doubles."""
    response = client.get(url, params=params, headers=headers)
    if inspect.isawaitable(response):
        return await response
    return response


def _first_present(data: dict[str, Any], key: str) -> Any:
    value = data.get(key)
    if isinstance(value, list):
        return value[0] if value else None
    return value


def _format_value(value: Any, unit: str | None, *, precision: int = 1) -> str | None:
    if value is None:
        return None
    if isinstance(value, int | float):
        text = f"{float(value):.{precision}f}".rstrip("0").rstrip(".")
    else:
        text = str(value).strip()
    if not text:
        return None
    return f"{text} {unit}".strip() if unit else text


def _format_direction(value: Any) -> str | None:
    if value is None:
        return None
    try:
        degrees = float(value)
    except (TypeError, ValueError):
        return str(value).strip() or None
    cardinal = degrees_to_cardinal(degrees)
    return f"{cardinal} ({degrees:.0f} degrees)"


def format_openmeteo_marine_report(
    data: dict[str, Any], location: Location
) -> SurfConditionReport | None:
    """Format Open-Meteo Marine API current data as accessible plain text."""
    current = data.get("current")
    if not isinstance(current, dict) or not current:
        return None
    units = data.get("current_units") if isinstance(data.get("current_units"), dict) else {}

    lines = [
        f"Surf conditions from Open-Meteo Marine for {location.name}.",
        "Marine/surf conditions from Open-Meteo Marine; not an official NWS Surf Zone Forecast.",
    ]

    fields = (
        (
            "Wave height",
            _format_value(_first_present(current, "wave_height"), units.get("wave_height")),
        ),
        ("Wave direction", _format_direction(_first_present(current, "wave_direction"))),
        (
            "Wave period",
            _format_value(_first_present(current, "wave_period"), units.get("wave_period")),
        ),
        (
            "Swell height",
            _format_value(
                _first_present(current, "swell_wave_height"), units.get("swell_wave_height")
            ),
        ),
        ("Swell direction", _format_direction(_first_present(current, "swell_wave_direction"))),
        (
            "Swell period",
            _format_value(
                _first_present(current, "swell_wave_period"), units.get("swell_wave_period")
            ),
        ),
        (
            "Sea surface temperature",
            _format_value(
                _first_present(current, "sea_surface_temperature"),
                units.get("sea_surface_temperature"),
            ),
        ),
    )
    for label, value in fields:
        if value:
            lines.append(f"{label}: {value}.")

    if len(lines) == 2:
        return None

    issued_at = None
    current_time = current.get("time")
    if isinstance(current_time, str) and current_time:
        try:
            issued_at = datetime.fromisoformat(current_time.replace("Z", "+00:00"))
        except ValueError:
            issued_at = None
    return SurfConditionReport(
        source_name="Open-Meteo Marine",
        product_id="openmeteo-marine-surf-conditions",
        issued_at=issued_at or datetime.now(UTC),
        text="\n".join(lines),
    )


async def fetch_openmeteo_marine_surf_conditions(
    location: Location,
    *,
    marine_base_url: str = OPENMETEO_MARINE_BASE_URL,
    client: httpx.AsyncClient | None = None,
    timeout: float = 10.0,
    user_agent: str = "AccessiWeather (github.com/orinks/accessiweather)",
) -> TextProduct | None:
    """Fetch a small Open-Meteo Marine surf/beach conditions summary."""
    if location.latitude is None or location.longitude is None:
        return None

    url = f"{marine_base_url}/marine"
    params = {
        "latitude": location.latitude,
        "longitude": location.longitude,
        "current": ",".join(_OPENMETEO_CURRENT_VARIABLES),
        "timezone": "auto",
    }
    headers = {"User-Agent": user_agent}

    async def _run(http_client: httpx.AsyncClient) -> TextProduct | None:
        try:
            response = await _client_get(http_client, url, params=params, headers=headers)
            if response.status_code != 200:
                logger.debug("Open-Meteo Marine returned HTTP %s", response.status_code)
                return None
            report = format_openmeteo_marine_report(response.json(), location)
            return report.to_text_product() if report is not None else None
        except (httpx.TimeoutException, httpx.TransportError, httpx.RequestError, ValueError):
            logger.debug("Open-Meteo Marine surf conditions unavailable", exc_info=True)
            return None

    if client is not None:
        return await _run(client)
    async with httpx.AsyncClient(timeout=timeout, follow_redirects=True) as new_client:
        return await _run(new_client)


async def fetch_pirate_weather_beach_conditions(
    location: Location,
    weather_client: object | None = None,
) -> TextProduct | None:
    """Build beach-weather context from an existing Pirate Weather client when available."""
    if weather_client is None:
        return None

    client_getter = getattr(weather_client, "_pirate_weather_client_for_location", None)
    pirate_client = client_getter(location) if callable(client_getter) else None
    if pirate_client is None:
        pirate_client = getattr(weather_client, "pirate_weather_client", None)
    if pirate_client is None:
        return None

    try:
        payload = await pirate_client.get_forecast_data(location)
    except Exception:  # noqa: BLE001
        logger.debug("Pirate Weather beach conditions unavailable", exc_info=True)
        return None
    if not isinstance(payload, dict):
        return None

    current = payload.get("currently")
    if not isinstance(current, dict):
        return None

    lines = [
        f"Surf conditions from Pirate Weather for {location.name}.",
        (
            "Beach-weather context from Pirate Weather; not an official NWS Surf Zone "
            "Forecast and wave data is not available from this source in AccessiWeather."
        ),
    ]
    for label, key, unit in (
        ("Conditions", "summary", ""),
        ("Temperature", "temperature", "degrees"),
        ("Feels like", "apparentTemperature", "degrees"),
        ("Wind speed", "windSpeed", "mph"),
        ("Wind gust", "windGust", "mph"),
        ("Wind direction", "windBearing", ""),
        ("UV index", "uvIndex", ""),
        ("Visibility", "visibility", "miles"),
    ):
        value = current.get(key)
        formatted = _format_direction(value) if key == "windBearing" else _format_value(value, unit)
        if formatted:
            lines.append(f"{label}: {formatted}.")

    precip_probability = current.get("precipProbability")
    if isinstance(precip_probability, int | float):
        lines.append(f"Precipitation chance: {precip_probability * 100:.0f} percent.")

    if len(lines) == 2:
        return None

    return TextProduct(
        product_type="SURF_CONDITIONS",
        product_id="pirate-weather-beach-conditions",
        cwa_office="Pirate Weather",
        issuance_time=datetime.now(UTC),
        product_text="\n".join(lines),
        headline="Surf conditions from Pirate Weather",
    )

File: rust/crates/aw-providers/src/surf_conditions.rs
//! Surf and beach-condition summaries, ported from
//! `accessiweather.surf_conditions` (Open-Meteo Marine plus Pirate Weather
//! beach-weather context, both shown as Forecaster Notes text products).

use aw_core::model::{Location, TextProduct, Timestamp};
use aw_core::py;
use aw_core::weather_client_parsers::degrees_to_cardinal;
use chrono::{Local, TimeZone, Utc};
use serde_json::Value;

use crate::http::{build_url, HttpClient};
use crate::pirateweather::PirateWeatherClient;

pub const OPENMETEO_MARINE_BASE_URL: &str = "https://marine-api.open-meteo.com/v1";
pub const DEFAULT_USER_AGENT: &str = "AccessiWeather (github.com/orinks/accessiweather)";
const CURRENT_VARIABLES: [&str; 7] = [
    "wave_height",
    "wave_direction",
    "wave_period",
    "swell_wave_height",
    "swell_wave_direction",
    "swell_wave_period",
    "sea_surface_temperature",
];

/// A source-labelled surf/beach conditions summary.
#[derive(Debug, Clone, PartialEq)]
pub struct SurfConditionReport {
    pub source_name: String,
    pub product_id: String,
    pub text: String,
    pub issued_at: Option<Timestamp>,
    pub official: bool,
}

impl SurfConditionReport {
    /// The Forecaster Notes text product shape.
    pub fn to_text_product(&self) -> TextProduct {
        TextProduct {
            product_type: "SURF_CONDITIONS".into(),
            product_id: self.product_id.clone(),
            cwa_office: self.source_name.clone(),
            issuance_time: self.issued_at,
            product_text: self.text.clone(),
            headline: Some(if self.official {
                format!("Official surf forecast from {}", self.source_name)
            } else {
                format!("Surf conditions from {}", self.source_name)
            }),
        }
    }
}

fn first_present<'a>(data: &'a Value, key: &str) -> Option<&'a Value> {
    match data.get(key)? {
        Value::Array(items) => items.first(),
        other => Some(other),
    }
    .filter(|v| !v.is_null())
}

fn format_value(value: Option<&Value>, unit: Option<&str>) -> Option<String> {
    let text = match value? {
        Value::Null => return None,
        v @ (Value::Number(_) | Value::Bool(_)) => {
            let formatted = format!("{:.1}", py::number(Some(v))?);
            formatted.trim_end_matches('0').trim_end_matches('.').to_string()
        }
        v => py::value_str(v).trim().to_string(),
    };
    if text.is_empty() {
        return None;
    }
    Some(match unit.filter(|u| !u.is_empty()) {
        Some(u) => format!("{text} {u}").trim().to_string(),
        None => text,
    })
}

fn format_direction(value: Option<&Value>) -> Option<String> {
    let value = value.filter(|v| !v.is_null())?;
    match py::as_float(Some(value)) {
        Some(degrees) => Some(format!(
            "{} ({degrees:.0} degrees)",
            degrees_to_cardinal(Some(degrees)).unwrap_or_default()
        )),
        None => Some(py::value_str(value).trim().to_string()).filter(|s| !s.is_empty()),
    }
}

/// Open-Meteo Marine current data as accessible plain text; `now` is the
/// fallback issue time. A naive `time` is read as local wall-clock time, as
/// Python does.
pub fn format_openmeteo_marine_report(
    data: &Value,
    location: &Location,
    now: Timestamp,
) -> Option<SurfConditionReport> {
    let current = data.get("current")?.as_object().filter(|c| !c.is_empty())?;
    let current = Value::Object(current.clone());
    let empty = serde_json::Map::new();
    let units = data.get("current_units").and_then(Value::as_object).unwrap_or(&empty);
    let unit = |key: &str| units.get(key).and_then(Value::as_str);

    let mut lines = vec![
        format!("Surf conditions from Open-Meteo Marine for {}.", location.name),
        "Marine/surf conditions from Open-Meteo Marine; not an official NWS Surf Zone Forecast."
            .to_string(),
    ];
    let fields = [
        ("Wave height", format_value(first_present(&current, "wave_height"), unit("wave_height"))),
        ("Wave direction", format_direction(first_present(&current, "wave_direction"))),
        ("Wave period", format_value(first_present(&current, "wave_period"), unit("wave_period"))),
        (
            "Swell height",
            format_value(first_present(&current, "swell_wave_height"), unit("swell_wave_height")),
        ),
        ("Swell direction", format_direction(first_present(&current, "swell_wave_direction"))),
        (
            "Swell period",
            format_value(first_present(&current, "swell_wave_period"), unit("swell_wave_period")),
        ),
        (
            "Sea surface temperature",
            format_value(
                first_present(&current, "sea_surface_temperature"),
                unit("sea_surface_temperature"),
            ),
        ),
    ];
    for (label, value) in fields {
        if let Some(value) = value.filter(|v| !v.is_empty()) {
            lines.push(format!("{label}: {value}."));
        }
    }
    if lines.len() == 2 {
        return None;
    }

    let issued_at = current
        .get("time")
        .and_then(Value::as_str)
        .filter(|s| !s.is_empty())
        .and_then(|s| py::fromisoformat(&s.replace('Z', "+00:00")))
        .and_then(|(naive, offset)| match offset {
            Some(off) => off.from_local_datetime(&naive).single(),
            None => Local.from_local_datetime(&naive).earliest().map(|t| t.fixed_offset()),
        });
    Some(SurfConditionReport {
        source_name: "Open-Meteo Marine".into(),
        product_id: "openmeteo-marine-surf-conditions".into(),
        issued_at: Some(issued_at.unwrap_or_else(|| now.to_utc().fixed_offset())),
        text: lines.join("\n"),
        official: false,
    })
}

/// Marine request URL.
pub fn marine_url(location: &Location, marine_base_url: &str) -> String {
    build_url(
        &format!("{marine_base_url}/marine"),
        &[
            ("latitude", py::float_repr(location.latitude)),
            ("longitude", py::float_repr(location.longitude)),
            ("current", CURRENT_VARIABLES.join(",")),
            ("timezone", "auto".into()),
        ],
    )
}

/// Fetch a small Open-Meteo Marine surf summary; any failure yields `None`.
pub fn fetch_openmeteo_marine_surf_conditions(
    http: &dyn HttpClient,
    location: &Location,
    marine_base_url: &str,
    user_agent: &str,
) -> Option<TextProduct> {
    let url = marine_url(location, marine_base_url);
    match http.get_json_with_headers(&url, &[("User-Agent", user_agent)]) {
        Ok(data) => {
            format_openmeteo_marine_report(&data, location, Utc::now().fixed_offset())
                .map(|r| r.to_text_product())
        }
        Err(e) => {
            tracing::debug!("Open-Meteo Marine surf conditions unavailable: {e}");
            None
        }
    }
}

/// Beach-weather context from a Pirate Weather payload (`now` = issue time).
pub fn pirate_weather_beach_conditions(
    payload: &Value,
    location: &Location,
    now: Timestamp,
) -> Option<TextProduct> {
    let current = payload.get("currently")?;
    if !current.is_object() {
        return None;
    }
    let mut lines = vec![
        format!("Surf conditions from Pirate Weather for {}.", location.name),
        "Beach-weather context from Pirate Weather; not an official NWS Surf Zone Forecast and \
         wave data is not available from this source in AccessiWeather."
            .to_string(),
    ];
    for (label, key, unit) in [
        ("Conditions", "summary", ""),
        ("Temperature", "temperature", "degrees"),
        ("Feels like", "apparentTemperature", "degrees"),
        ("Wind speed", "windSpeed", "mph"),
        ("Wind gust", "windGust", "mph"),
        ("Wind direction", "windBearing", ""),
        ("UV index", "uvIndex", ""),
        ("Visibility", "visibility", "miles"),
    ] {
        let value = current.get(key);
        let formatted = if key == "windBearing" {
            format_direction(value)
        } else {
            format_value(value, Some(unit))
        };
        if let Some(text) = formatted.filter(|t| !t.is_empty()) {
            lines.push(format!("{label}: {text}."));
        }
    }
    if let Some(p) = py::number(current.get("precipProbability")) {
        lines.push(format!("Precipitation chance: {:.0} percent.", p * 100.0));
    }
    if lines.len() == 2 {
        return None;
    }
    Some(TextProduct {
        product_type: "SURF_CONDITIONS".into(),
        product_id: "pirate-weather-beach-conditions".into(),
        cwa_office: "Pirate Weather".into(),
        issuance_time: Some(now.to_utc().fixed_offset()),
        product_text: lines.join("\n"),
        headline: Some("Surf conditions from Pirate Weather".into()),
    })
}

/// `fetch_pirate_weather_beach_conditions`: `None` without a configured client
/// or when the payload is unavailable.
pub fn fetch_pirate_weather_beach_conditions(
    location: &Location,
    pirate_client: Option<&PirateWeatherClient>,
) -> Option<TextProduct> {
    let payload = match pirate_client?.get_forecast_data(location) {
        Ok(p) => p,
        Err(e) => {
            tracing::debug!("Pirate Weather beach conditions unavailable: {e}");
            return None;
        }
    };
    pirate_weather_beach_conditions(&payload, location, Utc::now().fixed_offset())
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn now() -> Timestamp {
        chrono::DateTime::parse_from_rfc3339("2025-06-01T12:00:00+00:00").unwrap()
    }

    #[test]
    fn marine_report_labels_derived_conditions() {
        let data = json!({
            "current": {"time": "2025-06-01T08:00Z", "wave_height": 1.25, "wave_direction": 270,
                "wave_period": 9.0, "swell_wave_height": [0.8], "sea_surface_temperature": 21.46},
            "current_units": {"wave_height": "m", "wave_period": "s", "swell_wave_height": "m",
                "sea_surface_temperature": "°C"}
        });
        let report = format_openmeteo_marine_report(&data, &Location::new("Beach", 1.0, 2.0), now()).unwrap();
        assert_eq!(
            report.text,
            "Surf conditions from Open-Meteo Marine for Beach.\n\
             Marine/surf conditions from Open-Meteo Marine; not an official NWS Surf Zone Forecast.\n\
             Wave height: 1.2 m.\nWave direction: W (270 degrees).\nWave period: 9 s.\n\
             Swell height: 0.8 m.\nSea surface temperature: 21.5 °C."
        );
        let product = report.to_text_product();
        assert_eq!(product.headline.as_deref(), Some("Surf conditions from Open-Meteo Marine"));
        assert_eq!(product.issuance_time.unwrap().to_rfc3339(), "2025-06-01T08:00:00+00:00");
    }

    #[test]
    fn marine_report_requires_values() {
        let loc = Location::new("Beach", 1.0, 2.0);
        assert!(format_openmeteo_marine_report(&json!({"current": {"wave_height": null}}), &loc, now()).is_none());
        assert!(format_openmeteo_marine_report(&json!({"current": {}}), &loc, now()).is_none());
        let r = format_openmeteo_marine_report(
            &json!({"current": {"time": "bad", "wave_height": 1}}),
            &loc,
            now(),
        )
        .unwrap();
        assert_eq!(r.issued_at, Some(now()));
    }

    #[test]
    fn pirate_beach_conditions_use_available_context() {
        let payload = json!({"currently": {"summary": "Clear", "temperature": 78.44,
            "windSpeed": 10, "windBearing": 180, "precipProbability": 0.25}});
        let p = pirate_weather_beach_conditions(&payload, &Location::new("Beach", 1.0, 2.0), now()).unwrap();
        assert_eq!(
            p.product_text,
            "Surf conditions from Pirate Weather for Beach.\n\
             Beach-weather context from Pirate Weather; not an official NWS Surf Zone Forecast and \
             wave data is not available from this source in AccessiWeather.\n\
             Conditions: Clear.\nTemperature: 78.4 degrees.\nWind speed: 10 mph.\n\
             Wind direction: S (180 degrees).\nPrecipitation chance: 25 percent."
        );
        assert!(pirate_weather_beach_conditions(&json!({"currently": {}}), &Location::new("B", 0.0, 0.0), now()).is_none());
        assert!(fetch_pirate_weather_beach_conditions(&Location::new("B", 0.0, 0.0), None).is_none());
    }
}
